### ISSUE TITLE
[flang] static bounds checking of unreachable code

### ISSUE THREAD

**Issue by bcornille** [llvm/llvm-project#82684]
The below code encounters a semantic error check in flang-new. It may require some Fortran language lawyers to verify its correctness, but no runtime out of bounds access can occur in this program. Can someone please comment on whether this is truly a semantic error of the program or if flang is being too aggressive in its restriction? I encountered this pattern in a large application code and at first glance it seems like it would be a lot of work to use the preprocessor or other mechanism to generate code that flang would accept.

bounds_unreachable.f90:
```Fortran
program main
  implicit none
 
  integer, parameter :: N = ONE_OR_TWO
  integer :: array(N)
 
  array(1) = 1
  if (N > 1) then
    array(2) = 2
  endif
 
  write(*,*) array
 
end program main
```

Compilation error:
```shell
>> flang-new -cpp -DONE_OR_TWO=1 bounds_unreachable.f90
error: Semantic errors in bounds_unreachable.f90
./bounds_unreachable.f90:9:5: error: Subscript 2 is greater than upper bound 1 for dimension 1 of array
      array(2) = 2
      ^^^^^^^^
./bounds_unreachable.f90:5:14: Declaration of 'array'
    integer :: array(N)
               ^^^^^
```

gfortran behavior:
```shell
>> gfortran -cpp -DONE_OR_TWO=1 bounds_unreachable.f90
bounds_unreachable.f90:9:10:

     array(2) = 2
          1
Warning: Array reference at (1) is out of bounds (2 > 1) in dimension 1
```

**Comment by klausler** [comment]
The error is valid.

Change the code to `array(N) = 2` to avoid problems.

**Comment by bcornille** [comment]
Can you please provide reference to the Fortran specification that renders this code invalid?

**Comment by klausler** [comment]
Fortran 2023 subclause 9.5.3.1 paragraph 2.

**Comment by mjklemm** [comment]
I would argue that it is OK to emit a warning that can be optimally be turned into an error. There are big codes out there that will fail to compile because of this. Yes, it's UB code, but we might want to be a bit more forgiving.

**Comment by kkwli** [comment]
> I would argue that it is OK to emit a warning that can be optimally be turned into an error. There are big codes out there that will fail to compile because of this. Yes, it's UB code, but we might want to be a bit more forgiving.

I agree. A warning message instead of an error is more desirable. As @mjklemm, some old code out there will fail to compile.

**Comment by mjklemm** [comment]
I'll craft a PR for this next week.

**Comment by mjklemm** [comment]
PR #83011
